### PR TITLE
Add extension check against use of while loops

### LINKED ISSUE #4367: New check: Ban while-loops
The `while` operator is needed to make Python Turing-complete. But
unless you are implementing something like a theorem prover or an
interpreter, there is very little call to use a true `while` loop. In
practically every case a `while` loop can be converted to a `for` loop
with improved clarity. Seriously, how many web apps have a need for
unbounded computation?

One exception to this is the use of `while` loops at the tippity-top
of applications as a main execution loop. But other than that, genuine
need for `while` is rare.

I would like to see a check that simply flags any use of `while`. This
is an opinionated check, so make it an extension.

### PR DESCRIPTION
**Pull request by nickdrozd**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Description

Closes #4367

### CODE REVIEW

File: pylint/reporters/ureports/__init__.py [hunk 1]
```diff
@@ -75,7 +75,7 @@ def get_table_content(self, table):
             cols -= 1
             result[-1].append(cell)
         # fill missing cells
-        while len(result[-1]) < cols:
+        for _ in range(cols - len(result[-1])):
             result[-1].append("")
```
Comment (Pierre-Sassoulas) [inline pylint/reporters/ureports/__init__.py:79]:
```suggestion
        result[-1].append("" * (cols - len(result[-1])))
```

File: pylint/checkers/utils.py [hunk 2]
```diff
@@ -671,7 +671,7 @@ def node_frame_class(node: astroid.node_classes.NodeNG) -> Optional[astroid.Clas
         and not isinstance(klass, astroid.ClassDef)
     ):
         if klass.parent is None:
-            klass = None
+            return None
         else:
             klass = klass.parent.frame()
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/utils.py:676]:
```suggestion
        klass = klass.parent.frame()
```

File: pylint/extensions/typing.py [hunk 3]
```diff
@@ -293,11 +293,7 @@ def leave_module(self, node: astroid.Module) -> None:
         """
         if self._py37_plus() and not self._py39_plus():
             msg_future_import = self._msg_postponed_eval_hint(node)
-            while True:
-                try:
-                    msg = self._consider_using_alias_msgs.pop(0)
-                except IndexError:
-                    break
+            for msg in self._consider_using_alias_msgs:
```
Comment (Pierre-Sassoulas) [inline pylint/extensions/typing.py:296]: Wow, this one was particularly bad, good job fixing it 👌

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Thank you, this is pretty straightforward and the fixes to the existing code are on point too. Could you add an entry in what's new please ?

File: pylint/extensions/while.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+        self.add_message("while-used", node=node)
+
+
+def register(linter):
+    """Required method to auto register this checker.
+
+    :param linter: Main interface object for Pylint plugins
+    :type linter: Pylint object
+    """
+    linter.register_checker(WhileChecker(linter))
```
Comment (Pierre-Sassoulas) [inline pylint/extensions/while.py:30]: We should not register it, or it become a default checker and we don't want that :) But we can add a functional test where the plugin is loaded to check it.

File: pylint/reporters/ureports/__init__.py (hunk 1 above)
Comment (nickdrozd) [reply pylint/reporters/ureports/__init__.py:79]: Should be `result[-1] += "" * (cols - len(result[-1])`, right?

File: pylint/extensions/while.py (hunk 4 above)
Comment (nickdrozd) [reply pylint/extensions/while.py:30]: I think the fact that it is in `extensions` keeps it from being a default.

In any case, removing the register line causes `AttributeError: module 'pylint.extensions.while_used' has no attribute 'register'`.

File: pylint/reporters/ureports/__init__.py (hunk 1 above)
Comment (Pierre-Sassoulas) [reply pylint/reporters/ureports/__init__.py:79]: Hmm, nice catch, it's actually ``result[-1] += [""] * (cols - len(result[-1]))``

File: pylint/extensions/while.py (hunk 4 above)
Comment (Pierre-Sassoulas) [reply pylint/extensions/while.py]: Sorry, my bad :)

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Could you also add two functional tests, one  with the plugin activated, the other without configuration, please ?

**Comment by nickdrozd** [comment]
> Could you also add two functional tests, one with the plugin activated, the other without configuration, please ?

I added a functional test with the plugin activated, but I'm not sure what you mean about the other one. There are a variety of `while` loops used throughout the other functional tests and none of them are flagged. Is that what you had in mind?

**Review by Pierre-Sassoulas — APPROVED** [review]
Thank you for adding the test :) You're right, there's no need for another test, I did not think of that. This will be available in pylint 2.10.0 !

**Pull request merged**